README


This archive contains the distribution App-Scaffolder-Puppet,
version 0.003000:

  App::Scaffolder extension to scaffold Puppet modules

This software is copyright (c) 2017 by Manfred Stock.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for App-Scaffolder-Puppet

{{$NEXT}}

0.004000  2021-01-10 18:10:57+01:00 Europe/Zurich

 - Add type annotations to the parameters of the generated classes.
 - Use the Puppet Strings documentation format for the generated documentation.

0.003001  2017-07-16 17:08:04+02:00 Europe/Zurich

 - Add workaround for catdir() behaviour difference between Cygwin and Unix.

0.003000  2017-07-13 21:15:42+02:00 Europe/Zurich

 - Remove explicit hiera() calls, they are not required by current Puppet
   versions anymore.
 - Inherit from the parameter class in all sub-classes. Should make it a little
   easier to reuse them in another setting.

0.002001  2015-02-28 14:35:04 Europe/Zurich

 - Fix tests on Windows

0.002000  2015-02-26 20:59:30 Europe/Zurich

 - Add 'ensure' parameter to 'package' and 'service' templates
 - Add --package option to provide package name, use package name where it
   seems reasonable in existing templates
 - Add 'puppetclass' command to create single or related classes which don't
   constitute a complete Puppet module
 - Add 'subpackage' template which can be used with the 'puppetclass' command
 - Enhance some methods to better support the new 'subpackage' template
 - Change template search path to include the command name (so if templates
   were put in eg. /etc/puppet/scaffolder_templates, they will have to be moved
   to /etc/puppet/scaffolder_templates/{puppetmodule,puppetclass} now)

0.001001  2014-02-01 16:54:57 Europe/Zurich

 - Add --quiet option in test to avoid output
 - Make tests more robust on platforms that don't necessarily use '/' as
   directory/file separator

0.001000  2014-01-18 16:25:52 Europe/Zurich

 - Initial release


t/lib/App/Scaffolder/Puppet/CommandTest.pm
package App::Scaffolder::Puppet::CommandTest;
use parent qw(App::Scaffolder::Puppet::TestBase);

use strict;
use warnings;

use Carp;
use Test::More;
use Test::Exception;
use File::Spec::Functions qw(catfile);
use App::Scaffolder::Puppet::Command;

sub get_extra_template_dirs_test : Test(2) {
	my ($self) = @_;

	my $sharedir = Path::Class::Dir->new(qw(share));
	my $cmd = App::Scaffolder::Puppet::Command->new({});
	local $ENV{SCAFFOLDER_TEMPLATE_PATH} = $sharedir->stringify();
	my @template_dirs = $cmd->get_extra_template_dirs('puppetmodule');
	cmp_ok(scalar @template_dirs, '>=', 1, 'at least one template dir found');
	is(
		scalar(grep { $_ eq $sharedir->subdir('puppetmodule')->stringify() } @template_dirs),
		1,
		'path from environment respected'
	);
}


sub get_target_test : Test(3) {
	my ($self) = @_;

	my $cmd = App::Scaffolder::Puppet::Command->new({});

	$self->{name_opt} = 'foo';
	is($cmd->get_target($self->{opt_mock}), 'foo', 'target ok');

	$self->{name_opt} = 'foo::bar';
	is($cmd->get_target($self->{opt_mock}), '.', 'target ok');

	$self->{target_opt} = 'test/dir';
	is($cmd->get_target($self->{opt_mock}), 'test/dir', 'target parameter overrides');
}


sub get_variables_test : Test(3) {
	my ($self) = @_;

	my $cmd = App::Scaffolder::Puppet::Command->new({});

	$self->{name_opt} = 'foo';
	is_deeply($cmd->get_variables($self->{opt_mock}), {
		name               => 'foo',
		package            => 'foo',
		nameparts          => ['foo'],
		namepartsjoined    => 'foo',
		namepartspath      => 'foo',
		subnameparts       => [],
		subnamepartsjoined => '',
		subnamepartspath   => undef,
	}, 'variables ok');

	$self->{name_opt} = 'foo::bar';
	is_deeply($cmd->get_variables($self->{opt_mock}), {
		name               => 'foo::bar',
		package            => 'foo::bar',
		nameparts          => ['foo', 'bar'],
		namepartsjoined    => 'foo_bar',
		namepartspath      => catfile('foo', 'bar'),
		subnameparts       => ['bar'],
		subnamepartsjoined => 'bar',
		subnamepartspath   => 'bar',
	}, 'variables ok');

	$self->{name_opt} = 'foo::bar::baz';
	$self->{package_opt} = 'foo-bar-baz';
	is_deeply($cmd->get_variables($self->{opt_mock}), {
		name               => 'foo::bar::baz',
		package            => 'foo-bar-baz',
		nameparts          => ['foo', 'bar', 'baz'],
		namepartsjoined    => 'foo_bar_baz',
		namepartspath      => catfile('foo', 'bar', 'baz'),
		subnameparts       => ['bar', 'baz'],
		subnamepartsjoined => 'bar_baz',
		subnamepartspath   => catfile('bar', 'baz'),
	}, 'variables ok');
}


1;


t/lib/App/Scaffolder/CommandTest/puppetmoduleTest.pm
package App::Scaffolder::CommandTest::puppetmoduleTest;
use parent qw(Test::Class);

use strict;
use warnings;

use Carp;
use Test::More;
use Test::Exception;
use Test::MockObject;
use App::Cmd::Tester;
use Path::Class::Dir;
use File::Spec::Functions qw(catfile);
use Test::File;
use Test::File::ShareDir '-share' => {
	'-dist' => { 'App-Scaffolder-Puppet' => Path::Class::Dir->new(qw(share)) }
};
use Directory::Scratch;
use App::Scaffolder;
use App::Scaffolder::Command::puppetmodule;


sub setup : Test(setup) {
	my ($self) = @_;

	$self->{opt_mock} = Test::MockObject->new();
	delete $self->{target_opt};
	delete $self->{name_opt};
	delete $self->{package_opt};
	$self->{opt_mock}->mock(target => sub {
		return $self->{target_opt};
	});
	$self->{opt_mock}->mock(name => sub {
		return $self->{name_opt};
	});
	$self->{opt_mock}->mock(package => sub {
		return $self->{package_opt};
	});
}


sub app_test : Test(7) {
	my ($self) = @_;

	my $scratch = Directory::Scratch->new();
	my $result = test_app('App::Scaffolder' => [
		qw(puppetmodule --name vim --template package --quiet --target), $scratch->base()
	]);
	is($result->stdout(), '', 'no output');
	is($result->error, undef, 'threw no exceptions');
	my @files = (
		[qw(manifests init.pp)],
		[qw(manifests install.pp)],
		[qw(manifests params.pp)],
		[qw(tests init.pp)],
		[qw(.gitignore)],
	);
	for my $file (@files) {
		file_exists_ok($scratch->base()->file(@{$file}));
	}
}


sub get_extra_template_dirs_test : Test(2) {
	my ($self) = @_;

	my $sharedir = Path::Class::Dir->new(qw(share));
	my $cmd = App::Scaffolder::Command::puppetmodule->new({});
	local $ENV{SCAFFOLDER_TEMPLATE_PATH} = $sharedir->stringify();
	my @template_dirs = $cmd->get_extra_template_dirs('puppetmodule');
	cmp_ok(scalar @template_dirs, '>=', 1, 'at least one template dir found');
	is(
		scalar(grep { $_ eq $sharedir->subdir('puppetmodule')->stringify() } @template_dirs),
		1,
		'path from environment respected'
	);
}


sub get_target_test : Test(3) {
	my ($self) = @_;

	my $cmd = App::Scaffolder::Command::puppetmodule->new({});

	$self->{name_opt} = 'foo';
	is($cmd->get_target($self->{opt_mock}), 'foo', 'target ok');

	$self->{name_opt} = 'foo::bar';
	is($cmd->get_target($self->{opt_mock}), '.', 'target ok');

	$self->{target_opt} = 'test/dir';
	is($cmd->get_target($self->{opt_mock}), 'test/dir', 'target parameter overrides');
}


sub get_variables_test : Test(3) {
	my ($self) = @_;

	my $cmd = App::Scaffolder::Command::puppetmodule->new({});

	$self->{name_opt} = 'foo';
	is_deeply($cmd->get_variables($self->{opt_mock}), {
		name               => 'foo',
		package            => 'foo',
		nameparts          => ['foo'],
		namepartsjoined    => 'foo',
		namepartspath      => 'foo',
		subnameparts       => [],
		subnamepartsjoined => '',
		subnamepartspath   => undef,
	}, 'variables ok');

	$self->{name_opt} = 'foo::bar';
	is_deeply($cmd->get_variables($self->{opt_mock}), {
		name               => 'foo::bar',
		package            => 'foo::bar',
		nameparts          => ['foo', 'bar'],
		namepartsjoined    => 'foo_bar',
		namepartspath      => catfile('foo', 'bar'),
		subnameparts       => ['bar'],
		subnamepartsjoined => 'bar',
		subnamepartspath   => 'bar',
	}, 'variables ok');

	$self->{name_opt} = 'foo::bar::baz';
	$self->{package_opt} = 'foo-bar-baz';
	is_deeply($cmd->get_variables($self->{opt_mock}), {
		name               => 'foo::bar::baz',
		package            => 'foo-bar-baz',
		nameparts          => ['foo', 'bar', 'baz'],
		namepartsjoined    => 'foo_bar_baz',
		namepartspath      => catfile('foo', 'bar', 'baz'),
		subnameparts       => ['bar', 'baz'],
		subnamepartsjoined => 'bar_baz',
		subnamepartspath   => catfile('bar', 'baz'),
	}, 'variables ok');
}

1;


lib/App/Scaffolder/Puppet.pm
package App::Scaffolder::Puppet;
$App::Scaffolder::Puppet::VERSION = '0.003000';
# ABSTRACT: App::Scaffolder extension to scaffold Puppet modules

use strict;
use warnings;


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::Scaffolder::Puppet - App::Scaffolder extension to scaffold Puppet modules

=head1 VERSION

version 0.003000

=head1 DESCRIPTION

App::Scaffolder::Puppet provides commands to scaffold Puppet modules and classes.
See L<App::Scaffolder::Command::puppetmodule|App::Scaffolder::Command::puppetmodule>
and L<App::Scaffolder::Command::puppetclass|App::Scaffolder::Command::puppetclass>
for the actual commands.

L<App::Scaffolder::Puppet::Command|App::Scaffolder::Puppet::Command> is a base
class for the above commands, and is itself based on
L<App::Scaffolder::Command|App::Scaffolder::Command>.

=head1 SEE ALSO

=over

=item *

L<App::Scaffolder|App::Scaffolder>

=item *

L<App::Scaffolder::Puppet::Command|App::Scaffolder::Puppet::Command>

=item *

L<https://puppetlabs.com/puppet/puppet-open-source> - Puppet

=back

=head1 AUTHOR

Manfred Stock <mstock@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Manfred Stock.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/App/Scaffolder/Puppet/Command.pm
package App::Scaffolder::Puppet::Command;

use parent qw(App::Scaffolder::Command);

# ABSTRACT: Base class for App::Scaffolder::Puppet commands

use strict;
use warnings;

use File::Spec::Functions qw(catdir);
use MRO::Compat;
use Path::Class::Dir;

=head1 SYNOPSIS

	use parent qw(App::Scaffolder::Puppet::Command);

=head1 DESCRIPTION

This class provides some specialized methods for the C<puppet*> commands.

In addition to the default template search path (see
L<App::Scaffolder|App::Scaffolder> for details), this command base class will
also look for templates in C</etc/puppet/scaffolder_templates> or
C</usr/local/etc/puppet/scaffolder_templates> if they exist.

=head1 METHODS

=head2 get_target

Specialized C<get_target> version which uses the name (if it does not contain
C<::>, otherwise, it will be the current working directory) if no target was
given.

=cut

sub get_target {
	my ($self, $opt) = @_;
	my $target = $opt->target() || (
		$opt->name() =~ m{::}x
			? '.'
			: $opt->name()
	);
	return Path::Class::Dir->new($target);
}


=head2 get_variables

Specialized C<get_variables> version which returns the name of the module and
other useful variables.

=cut

sub get_variables {
	my ($self, $opt) = @_;

	my @name_parts = split(/::/x, $opt->name());
	my (undef, @subname_parts) = @name_parts;
	my $package = $opt->package() || $opt->name();
	return {
		name               => scalar $opt->name(),
		nameparts          => \@name_parts,
		namepartsjoined    => join('_', @name_parts),
		namepartspath      => catdir(@name_parts) || '',
		subnameparts       => \@subname_parts,
		subnamepartsjoined => join('_', @subname_parts),
		subnamepartspath   => catdir(@subname_parts) || '',
		package            => $package,
	};
}


=head2 get_dist_name

Return the name of the dist this command is in.

=cut

sub get_dist_name {
	return 'App-Scaffolder-Puppet';
}


=head2 get_options

Return additional options for this command.

=cut

sub get_options {
	my ($class) = @_;
	return (
		[ 'name|n=s'    => 'Name of the new Puppet module/class that should be created' ],
		[ 'package|p=s' => 'Name of a package that should be available in templates '
			. '(defaults to the value of the --name parameter)' ],
	);
}


sub validate_args {
	my ($self, $opt, $args) = @_;

	$self->next::method($opt, $args);
	unless ($self->contains_base_args($opt) || $opt->name()) {
		$self->usage_error("Parameter 'name' required");
	}
	return;
}


=head2 get_extra_template_dirs

Extend the template search path with
C</etc/puppet/scaffolder_templates/E<lt>command nameE<gt>> or
C</usr/local/etc/puppet/scaffolder_templates/E<lt>command nameE<gt>> if they
exist.

=head3 Result

The extended list with template directories.

=cut

sub get_extra_template_dirs {
	my ($self, $command) = @_;

	my @sub_path = ('etc', 'puppet', 'scaffolder_templates', $command);
	my @extra_template_dirs = grep { -d $_ && -r $_ } (
		Path::Class::Dir->new('', @sub_path),
		Path::Class::Dir->new('', 'usr', 'local', @sub_path),
	);

	return (
		$self->next::method($command),
		@extra_template_dirs,
	);
}


1;

=head1 SEE ALSO

=over

=item *

L<App::Scaffolder::Command|App::Scaffolder::Command>

=back


lib/App/Scaffolder/Command/puppetclass.pm
package App::Scaffolder::Command::puppetclass;
$App::Scaffolder::Command::puppetclass::VERSION = '0.003000';
use parent qw(App::Scaffolder::Puppet::Command);

# ABSTRACT: Scaffold one or more related Puppet classes

use strict;
use warnings;


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::Scaffolder::Command::puppetclass - Scaffold one or more related Puppet classes

=head1 VERSION

version 0.003000

=head1 SYNOPSIS

	# Create scaffold to install the 'vim-puppet' package in module created with
	# puppetmodule command and 'package' template:
	$ scaffolder puppetclass --template subpackage --name vim::puppet --package vim-puppet

	# Create scaffold to install the 'apache2-doc' package in module created with
	# puppetmodule command and 'service' template:
	$ scaffolder puppetclass --template subpackage --name apache2::doc --package apache2-doc

=head1 DESCRIPTION

App::Scaffolder::Command::puppetclass scaffolds one or more related Puppet
classes. It does not create a complete Puppet module (see
L<App::Scaffolder::Command::puppetmodule|App::Scaffolder::Command::puppetmodule>
for this), it just adds additional (usually closely related) classes to an
existing module. By default, it provides the following templates:

=over

=item *

C<subpackage>: Create class to install a 'sub package'. This is intended to be
used after using the C<package> or C<service> templates of the
L<App::Scaffolder::Command::puppetmodule|App::Scaffolder::Command::puppetmodule>
command to add an additional package to the module (eg. C<apache2-doc> to the
C<apache2> service). This must be used inside the module directory created before,
and you will have to add a variable with the actual package name to the existing
C<manifests/params.pp> file. The name of the variable can be seen in the newly
created files below C<manifests>.

=back

=head1 SEE ALSO

=over

=item *

L<App::Scaffolder::Command::puppetmodule|App::Scaffolder::Command::puppetmodule>

=back

=head1 AUTHOR

Manfred Stock <mstock@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Manfred Stock.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/App/Scaffolder/Command/puppetmodule.pm
package App::Scaffolder::Command::puppetmodule;
$App::Scaffolder::Command::puppetmodule::VERSION = '0.003000';
use parent qw(App::Scaffolder::Puppet::Command);

# ABSTRACT: Scaffold a Puppet module

use strict;
use warnings;


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::Scaffolder::Command::puppetmodule - Scaffold a Puppet module

=head1 VERSION

version 0.003000

=head1 SYNOPSIS

	# Create scaffold to install the 'vim' package:
	$ scaffolder puppetmodule --template package --name vim

	# Create scaffold to install the 'apache2' package and setup the corresponding service:
	$ scaffolder puppetmodule --template service --name apache2

=head1 DESCRIPTION

App::Scaffolder::Command::puppetmodule scaffolds Puppet modules. By default, it
provides the following templates:

=over

=item *

C<package>: Create Puppet module to install a package.

=item *

C<service>: Create Puppet module to setup a service.

=back

=head1 SEE ALSO

=over

=item *

L<https://docs.puppetlabs.com/puppet/latest/reference/modules_fundamentals.html> - Module Fundamentals

=item *

L<App::Scaffolder::Command::puppetclass|App::Scaffolder::Command::puppetclass>

=back

=head1 AUTHOR

Manfred Stock <mstock@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Manfred Stock.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "App::Scaffolder extension to scaffold Puppet modules",
  "dist_author" => [
    "Manfred Stock <mstock\@cpan.org>"
  ],
  "dist_name" => "App-Scaffolder-Puppet",
  "dist_version" => "0.003000",
  "license" => "perl",
  "module_name" => "App::Scaffolder::Puppet",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "App::Scaffolder" => "0.002000",
    "App::Scaffolder::Command" => 0,
    "File::Spec::Functions" => 0,
    "MRO::Compat" => 0,
    "Path::Class::Dir" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "share_dir" => {
    "dist" => "share"
  },
  "test_requires" => {
    "App::Cmd::Tester" => 0,
    "App::Scaffolder" => "0.002000",
    "Carp" => 0,
    "Directory::Scratch" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Test::Class" => 0,
    "Test::Exception" => 0,
    "Test::File" => 0,
    "Test::File::ShareDir" => 0,
    "Test::MockObject" => 0,
    "Test::More" => "0.88",
    "lib" => 0
  }
);


my %fallback_build_requires = (
  "App::Cmd::Tester" => 0,
  "App::Scaffolder" => "0.002000",
  "Carp" => 0,
  "Directory::Scratch" => 0,
  "File::Find" => 0,
  "File::Temp" => 0,
  "Module::Build" => "0.3601",
  "Test::Class" => 0,
  "Test::Exception" => 0,
  "Test::File" => 0,
  "Test::File::ShareDir" => 0,
  "Test::MockObject" => 0,
  "Test::More" => "0.88",
  "lib" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
